apollo-spaceport/src/server.rs

```diff
@@ -0,0 +1,328 @@
+                        }
+                    },
+                    _ = interval.tick() => {
+                        tracing::debug!("spaceport ticked");
+                        task_total.store(0, Ordering::SeqCst);
+                        process_all_graphs(&client, task_graph_usage.clone()).await;
+                    }
+                };
+            }
+        });
+        Self {
+            addr,
+            graph_usage,
+            tx,
+            total,
+        }
+    }
+
+    /// Start serving requests.
+    pub async fn serve(self) -> Result<(), Error> {
+        let addr = self.addr;
+        Server::builder()
+            .add_service(ReporterServer::new(self))
+            .serve(addr)
+            .await
+    }
+
+    async fn submit_report(
+        client: &Client,
+        key: String,
+        report: Report,
+    ) -> Result<Response<ReporterResponse>, Status> {
+        tracing::debug!("submitting report: {:?}", report);
+        // Protobuf encode message
+        let mut content = BytesMut::new();
+        report
+            .encode(&mut content)
+            .map_err(|e| Status::invalid_argument(e.to_string()))?;
+        // Create a gzip encoder
+        let mut encoder = GzEncoder::new(Vec::new(), Compression::default());
+        // Write our content to our encoder
+        encoder
+            .write_all(&content)
+            .map_err(|e| Status::internal(e.to_string()))?;
+        // Finish encoding and retrieve content
+        let compressed_content = encoder
+            .finish()
+            .map_err(|e| Status::internal(e.to_string()))?;
+        let mut backoff = Duration::from_millis(0);
+        let ingress = match std::env::var("APOLLO_USAGE_REPORTING_INGRESS_URL") {
+            Ok(v) => v,
+            Err(_e) => DEFAULT_APOLLO_USAGE_REPORTING_INGRESS_URL.to_string(),
+        };
+        let req = client
+            .post(ingress)
+            .body(compressed_content)
+            .header("X-Api-Key", key)
+            .header("Content-Encoding", "gzip")
+            .header("Content-Type", "application/protobuf")
+            .header("Accept", "application/json")
+            .build()
+            .map_err(|e| Status::unavailable(e.to_string()))?;
+
+        for i in 0..5 {
+            // We know these requests can be cloned
+            let task_req = req.try_clone().expect("requests must be clone-able");
+            match client.execute(task_req).await {
+                Ok(v) => {
+                    let data = v
+                        .text()
+                        .await
+                        .map_err(|e| Status::internal(e.to_string()))?;
+                    tracing::debug!("ingress response text: {:?}", data);
+                    let response = ReporterResponse {
+                        message: "Report accepted".to_string(),
+                    };
+                    return Ok(Response::new(response));
+                }
+                Err(e) => {
```

Comment:
What kind of error does this represent?  Not a reqwest expert and the docs aren't super clear but I think this is only low-level network errors, and not 4xx/5xx from server?  I think that it is only low-level errors, and you'll want the `Ok(v)` case to check `v.status()` and see if `is_client_error()` in which case you fail without retry, or `is_server_error()` in which case you fail with retry.

(BTW the docs say that you should use "RequestBuilder and RequestBuilder::send()" instead of execute. Dunno why. Maybe you need it because of your request reuse.)

Reply:
I'll file an issue to clean up the error handling here.

(I am using execute() because of request reuse.)